pikaur/version.py

```diff
@@ -1,5 +1,125 @@
 from distutils.version import LooseVersion
 import ctypes
+import re
+from itertools import zip_longest
+
+
+class PackageVersion:
+    VERSION_RE = re.compile(r'''
+        ((?P<epoch>\d*):)? # epoch?
+        (?P<pkgver>[^-]+)  # pkgver
+        (-(?P<pkgrel>.*))?   # pkgrel?
+        $
+    ''', re.VERBOSE)
+
+    PKGVER_PART_RE = re.compile(r'(\d+)|([a-zA-Z]+)')
+
+    def __init__(self, version_string):
+        self.version_string = version_string
+
+        match = self.VERSION_RE.match(version_string)
+        epoch = match.group('epoch')
+        if not epoch:
+            epoch = 0
+        else:
+            epoch = int(epoch)
+
+        self.epoch = epoch
+        self.pkgver = match.group('pkgver')
+        self.pkgrel = match.group('pkgrel')
+
+        if self.pkgrel == '':
+            self.pkgrel = None
+
+        self.pkgver_span = match.span('pkgver')
+
+        self.pkgver_parts = self._get_pkgver_parts(self.pkgver, offset=self.pkgver_span[0])
+
+
+    def _get_pkgver_parts(self, pkgver, offset):
+        parts = []
+        for match in self.PKGVER_PART_RE.finditer(pkgver):
+            if match[1] != None:
+                part = int(match[1])
+            else:
+                part = match[2]
+
+            span = match.span()
+            parts.append((part, (offset + span[0], offset + span[1])))
+
+        return parts
+
+    def compare(self, other):
```

Comment:
please use `pikaur.version.compare_versions` instead of reimplementing comparison again

Reply:
We might as well implement it as it doesn't add any complexity to the code and it can be used in tests to check that it is consistent with `vercmp`.

I think the best solution is to rename the function to `_compare` and add the following method:

```
def common_prefix(self, other):
    _, prefix = self._compare(other)
    return self.version_string[:prefix]
```

and use that. We can then use `_compare` in the test code.

Reply:
i think it will be better to keep as simple as possible, so implement only a dedicated function for version diff, but not to do any extra stuff

Reply:
Literally the only difference to the code would be that `compare` would only return one value instead of a pair. We would still need exactly the same number of lines in `compare`.

How about doing what I said above, but also remove the `__str__` and `__lt__`, `__eq__`, etc. methods?

Reply:
so i would like to see that PR to be much smaller with just one new function for getting common part of the string, because less code always better in both terms of support and mainainability  -- i am more happy to merge PRs which have more red lines than green ones ;)

Reply:
The problem is that the common string between 2 versions isn't just a common prefix.

If the first version is `1.2.3-4` and the second one is `1.02.003-5` then the prefixes would be

```
[1.2.3]-4    =>    [1.02.003]-5
```

We can of course still do this with a single version, but then we would need to parse the two version strings twice so in total 4 parsings instead of the 2 we do currently.

Reply:
i think we need just to compare splitted versions by equality of blocks

Reply:
You forgot `:`, but we might as well do it properly like `pacman`/`vercmp` does it.

Reply:
in that case instead of just comparing them for equality we can pass each block to `compare_versions` func, ie to vercmp

Reply:
That wouldn't actually work as `1.2` and `1.2-5` should compare equal, whereas `1.2` and `1.2.5` shouldn't.

Also `1:2.3` and `1.2.3` shouldn't compare equal either.

Reply:
i think it should be enough just to additionally compare separator itself for equality

Reply:
I don't understand why we can't just do it correctly.

Do I really have to fork `pikaur` if I want correct version diffing behavior?

Reply:
tried my suggestion:

![2018-02-25--1519531649_1327x1888_scrot](https://user-images.githubusercontent.com/1655669/36638055-ce097656-19e9-11e8-9223-b71873a8e472.png)

Reply:
whole patch is that small:

```diff
diff --git a/pikaur/pprint.py b/pikaur/pprint.py
index 2168f95..c6958c3 100644
--- a/pikaur/pprint.py
+++ b/pikaur/pprint.py
@@ -2,6 +2,7 @@ import sys
 import shutil
 
 from .config import VERSION
+from .version import compare_versions
 
 
 PADDING = 4
@@ -61,6 +62,9 @@ def print_not_found_packages(not_found_packages):
         print(format_paragraph(package))
 
 
+VERSION_SEPARATORS = ('.', '+', '-', ':')
+
+
 def pretty_format_upgradeable(packages_updates, verbose=False, print_repo=False, color=True):
 
     _color_line = color_line
@@ -68,21 +72,32 @@ def pretty_format_upgradeable(packages_updates, verbose=False, print_repo=False,
     if not color:
         _color_line = _bold_line = lambda x, *args: x
 
-    def get_common_string(str1, str2):
-        result = ''
-        if '' in (str1, str2):
-            return result
-        counter = 0
-        while (
-                counter < len(str1)
-        ) and (
-            counter < len(str2)
-        ) and (
-            str1[counter] == str2[counter]
-        ):
-            result += str1[counter]
-            counter += 1
-        return result
+    def split_version(version):
+        splitted_version = []
+        block = ''
+        for char in version:
+            if char in VERSION_SEPARATORS:
+                splitted_version.append(block)
+                splitted_version.append(char)
+                block = ''
+            else:
+                block += char
+        return splitted_version
+
+    def get_common_version(version1, version2):
+        common_string = ''
+        common_length = 0
+        if '' in (version1, version2):
+            return common_string, common_length
+        splitted_version1 = split_version(version1)
+        splitted_version2 = split_version(version2)
+        for block1, block2 in zip(splitted_version1, splitted_version2):
+            if compare_versions(block1, block2) == 0:
+                common_string += block1
+                common_length += 1
+            else:
+                break
+        return common_string, common_length
 
     def get_version_diff(version, common_version):
         new_version_postfix = version
@@ -94,7 +109,7 @@ def pretty_format_upgradeable(packages_updates, verbose=False, print_repo=False,
         return new_version_postfix
 
     def pretty_format(pkg_update):
-        common_version = get_common_string(
+        common_version, difference_size = get_common_version(
             pkg_update.Current_Version, pkg_update.New_Version
         )
         version_color = 10
@@ -102,7 +117,7 @@ def pretty_format_upgradeable(packages_updates, verbose=False, print_repo=False,
         new_color = 9
         column_width = min(int(get_term_width() / 2.5), 37)
         sort_by = '{:03d}{}'.format(
-            len(common_version)*10+len(pkg_update.New_Version),
+            difference_size,
             pkg_update.Name
         )
         pkg_name = _bold_line(pkg_update.Name)
```

Reply:
is any of the diffs is looking incorrect on the example screenshot?

Reply:
It looks fine, however I would still prefer my solution.